---
title: "CRISPRi library V2, data processing pipeline"
output:
  html_notebook: 
    theme: cosmo
    toc: yes
    number_sections: true
---

----------

# Description

This R notebook details the data processing and visualization for growth competition experiments with a CRISPRi sgRNA library. The library contains around 20,000 unique sgRNA repression mutants tailored for the cyanobacterium _Synechocystis_ sp. PCC6803. This library is the second version (therefore "V2") of an sgRNA library for _Synechocystis_, containing 5 instead of only 2 sgRNAs per gene if possible. Some genes or ncRNAs are very short and hence it was no possible to design the maximum number of individual sgRNAs. The first iteration of the library was [published in Nature Communications, 2020](https://www.nature.com/articles/s41467-020-15491-7).

# Quality control

## Data import

Load required packages.

```{r, message = FALSE }
suppressPackageStartupMessages({
  library(tidyverse)
  library(ggrepel)
  library(lattice)
  library(latticeExtra)
  library(latticetools)
  library(scales)
  library(dendextend)
  library(vegan)
  library(tsne)
  library(KEGGREST)
  library(limma)
  library(corrplot)
})
```

Load raw data. The main table contains already normalized quantification of all sgRNAs, fold change, multiple hypothesis corrected p-values, and fitness score. Contrary to the processing of [our first CRISPRi library V1](https://github.com/m-jahn/R-notebooks), much of the functionality from the notebook was transferred into the [new CRISPRi library pipeline on github](https://github.com/m-jahn/CRISPRi-lib-pipe).

```{r}
load("../data/input/DESeq2_result.Rdata", )
df_main <- DESeq_result_table; rm(DESeq_result_table)
```


```{r, echo = FALSE}
# custom ggplot2 theme that is reused for all later plots
custom_colors = c("#E7298A", "#66A61E", "#E6AB02", "#7570B3", "#666666", "#1B9E77", "#D95F02", "#A6761D")
custom_theme <- function(base_size = 12, base_line_size = 1.0, base_rect_size = 1.0, ...) {
  theme_light(base_size = base_size, base_line_size = base_line_size, base_rect_size = base_rect_size) + theme(
    title = element_text(colour = grey(0.4), size = 10),
    plot.margin = unit(c(12,12,12,12), "points"),
    axis.ticks.length = unit(0.2, "cm"),
    axis.ticks = element_line(colour = grey(0.4), linetype = "solid", lineend = "round"),
    axis.text.x = element_text(colour = grey(0.4), size = 10),
    axis.text.y = element_text(colour = grey(0.4), size = 10),
    panel.grid.major = element_line(size = 0.6, linetype = "solid", colour = grey(0.9)),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(linetype = "solid", colour = grey(0.4), fill = NA, size = 1.0),
    panel.background = element_blank(),
    strip.background = element_blank(),
    strip.text = element_text(colour = grey(0.4), size = 10, margin = unit(rep(3,4), "points")),
    legend.text = element_text(colour = grey(0.4), size = 10),
    legend.title = element_blank(),
    legend.background = element_blank(),
    ...
  )
}
```


## Data annotation

Different annotation columns are added to the main data frame, including a short sgRNA identifier (excluding the position on the gene), an sgRNA index (1 to 5), and genome annotation from Uniprot. The Uniprot data is dynamically downloaded for every update of this pipeline using their very simple API (`read_tsv("https://www.uniprot.org/uniprot/?query=taxonomy:1111708&format=tab")`). The full list of columns that can be queried is available [here](https://www.uniprot.org/help/uniprotkb_column_names).
Pathway annotation from KEGG is later in the pipeline added using the `KEGGREST` package.

```{r, message = FALSE}
df_main <- df_main %>%
  # correct an error in sgRNA naming
  mutate(sgRNA = gsub('”', '2', sgRNA)) %>%
  # split sgRNA names into target gene and position
  separate(sgRNA, into = c("sgRNA_target", "sgRNA_position"), sep = "\\|",
    remove = FALSE) %>%
  
  # add sgRNA index number (1 to maximally 5) and type
  group_by(sgRNA_target) %>%
  mutate(
    sgRNA_index = sgRNA_position %>% as.numeric %>% as.factor %>%
    as.numeric,
    sgRNA_type = if_else(grepl("^nc_", sgRNA), "ncRNA", "gene")) %>%

  # map trivial names to LocusTags using a manually curated list
  left_join(
    read_tsv("../data/input/mapping_trivial_names.tsv", col_types = cols()),
    by = c("sgRNA_target" = "gene")) %>%
  
  # remove 2 conditions without response
  filter(!condition %in% c("BG11", "air, BG11, 200uE")) %>%
  mutate(condition = str_remove(condition, "BG11, ") %>%
    str_replace("air", "0.05% CO2")) %>%
  
  # remove some empty rows (NA targets)
  filter(!is.na(sgRNA_target))

head(df_main)
```

Optionally retrieve gene info from uniprot and merge with main data frame.

```{r}
df_uniprot <- read_tsv(paste0(
     "https://www.uniprot.org/uniprot/?query=taxonomy:1111708&format=tab&",
     "columns=id,genes,genes(PREFERRED),protein_names,length,mass,ec,database(KEGG)"),
     col_types = cols()) %>%
    rename_with(tolower) %>%
    rename(locus = `cross-reference (kegg)`, gene_name = `gene names`,
      gene_name_short = `gene names  (primary )`, ec_number = `ec number`,
      protein = `protein names`, uniprot_ID = entry
    ) %>%
  separate_rows(locus, sep = ";syn:") %>%
  mutate(locus = str_remove_all(locus, "syn:|;")) %>%
  filter(!is.na(locus))

df_main <- left_join(df_main, filter(df_uniprot, !duplicated(locus)),
  by = "locus")
```


## Number of sgRNAs

Each gene is represented by up to five sgRNAs. We can test if all or only some of the 5 sgRNAs are "behaving" in the same way in the same conditions, more mathematically speaking we can estimate the correlation of every sgRNA with another. First let's summarize how many genes have 5, 4, 3 sgRNAs and so on associated with them.

```{r, , fig.width = 6, fig.height = 3.5, message =FALSE}
# N unique sgRNAs in dataset
paste0("Number of unique sgRNAs: ", unique(df_main$sgRNA) %>% length)

# N genes with 1,2,3,4 or 5 sgRNAs
df_main %>%
  group_by(sgRNA_type, sgRNA_target) %>%
  summarize(n_sgRNAs = length(unique(sgRNA_position))) %>%
  count(n_sgRNAs) %>% filter(n_sgRNAs <= 5) %>%
  ggplot(aes(x = factor(n_sgRNAs), y = n, fill = factor(n_sgRNAs),
    label = n)) +
  geom_col(show.legend = FALSE) +
  geom_text(size = 3, nudge_y = 100) +
  facet_grid(~sgRNA_type) +
  custom_theme() +
  scale_fill_manual(values = custom_colors)
```

## Correlation of sgRNAs

Different methods can be used to estimate similarity between samples (sgRNAs). For example, factor analysis is a method to dissect underlying sources of variation within the dataset, and the contribution to overall variation. The most famous example is principal component analysis (PCA). More simple is to use the correlation coefficient of sgRNAs to each other to see if one of the sgRNAs contributes stronger to overall variation.

This is an example of an apparently strictly essential gene, encoding the ribosomal protein `rps10`. Most of the sgRNA repressor strains are depleted, the correlation between sgRNAs is high. The strength of depletion varies though, and the strain with sgRNA 3 is not depleted at all. We want to give higher weights to sgRNAs that correlate well with each other, and/or show stronger effect (depletion/enrichment).

```{r, fig.width = 7, fig.height = 4.5}
df_main %>% filter(sgRNA_target == "rps10") %>%
  mutate(sgRNA_index = factor(sgRNA_index)) %>%
  ggplot(aes(x = time, y = log2FoldChange, color = sgRNA_index)) +
  geom_line(size = 1) + geom_point(size = 2) +
  facet_wrap(~ condition, ncol = 3) +
  custom_theme() +
  scale_color_manual(values = custom_colors)
```

A correlation score can be calculated by computing the correlation coefficient of all sgRNAs to each other. This score is robustly summarized by taking the median, and rescaling it from the respective minima and maxima [-1, 1] to [0, 1]. This score serves as a weight for each sgRNA to calculate the (global) weighted mean of log2 FC over all sgRNAs. The score has the characteristic that it gives a weight of 1 for an sgRNA perfectly correlated with all other sgRNAs of the same gene, and a weight of 0 for sgRNAs perfectly anti-correlated to the other sgRNAs.

For a matrix of $x = 1 .. m$ sgRNAs and $y = 1 .. n$ observations (measurements), the correlation $R$ of one sgRNA to another is calculated using Pearson's method:

$R_x=cor([log_2FC_{x1,y1} ... log_2FC_{x1,yn}], [log_2FC_{x2,y1} ... log_2FC_{x2,yn}])$

The weight of one sgRNA is then calculated as median of all $R$ rescaled between 0 and 1.

$w_x = \frac{1 + median(R_1, R_2, ..., R_m)}{2}$

The following example shows the correlation matrix for the 5 `rps10` sgRNAs, and their weights. The self correlation of each sgRNA (R = 1) is removed prior to weight determination.

```{r, fig.width = 4, fig.height = 4}
cor_matrix <- df_main %>% filter(sgRNA_target == "rps10") %>% ungroup %>%
  select(sgRNA_index, log2FoldChange, group) %>%
  pivot_wider(names_from = group, values_from = log2FoldChange) %>%
  arrange(sgRNA_index) %>% column_to_rownames("sgRNA_index") %>%
  as.matrix %>% t %>% cor(method = "pearson")

weights <- cor_matrix %>% replace(., . == 1, NA) %>%
  apply(2, function(x) median(x, na.rm = TRUE)) %>%
  rescale(from = c(-1, 1), to = c(0, 1))

# plot heatmap
lattice::levelplot(cor_matrix %>% replace(., . == 1, NA),
  col.regions = colorRampPalette(custom_colors[c(1,5,2)])(20))

# print weights
weights
```

----------

Now we can create a function that will compute weights for all sgRNAs, and add the weights to the data set.

```{r, warning = FALSE}
determine_weight <- function(index, value, group) {
  # make correlation matrix
  df <- data.frame(index = index, value = value, group = group)
  cor_matrix <- pivot_wider(df, names_from = group, values_from = value) %>%
    arrange(index) %>% column_to_rownames("index") %>%
    as.matrix %>% t %>% cor(method = "pearson")
  
  # determine weights
  weights <- cor_matrix %>% replace(., . == 1, NA) %>%
    apply(2, function(x) median(x, na.rm = TRUE)) %>%
    scales::rescale(from = c(-1, 1), to = c(0, 1)) %>%
    enframe("index", "weight") %>% mutate(index = as.numeric(index)) %>%
    mutate(weight = replace(weight, is.na(weight), 1))
  
  # return vector of weights the same order and length 
  # as sgRNA index vector
  left_join(df, weights, by = "index") %>% pull(weight)
}

df_main <- df_main %>%
  group_by(sgRNA_target) %>%
  mutate(sgRNA_weight = determine_weight(sgRNA_index, log2FoldChange, group))
```


## Efficiency of sgRNAs

The correlation of each sgRNA with each other is a "global" parameter as it is identical over all conditions. A second global parameter, **sgRNA efficiency**, can be obtained using a similar approach. We expect that fitness of all sgRNAs for one gene is not normally distributed because sgRNAs are not ideal replicate measurements. They are biased by position effects and off-target binding, see [Wang et al., Nature Comms, 2018](https://www.nature.com/articles/s41467-018-04899-x) for a very insightful and comprehensive analysis of the number and position of sgRNAs required to estimate gene fitness. 

We calculate sgRNA efficiency $E$ as the median absolute fitness (AUC of log2FC over time) of an sgRNA $x = 1 .. m$ over all observations [conditions] $y = 1 .. n$.

$E_x=median(abs(fitness_{x1, y1}, fitness_{x1, y2}, ..., fitness_{x1, yn}))$

To normalize between all sgRNAs, $E$ is rescaled to a range between 0 and 1.

$E_x=\frac{E_x}{max(E_1, E_2, ..., E_m)}$

```{r}
df_main <- df_main %>% group_by(sgRNA_target) %>%
  mutate(sgRNA_efficiency = ave(fitness, sgRNA_index, FUN = function(x) median(abs(x))) %>%
    {./max(.)})
```

This is the resulting sgRNA efficiency for the example gene above, `rps10`.

```{r}
df_main %>% filter(sgRNA_target == "rps10") %>% ungroup %>%
  select(sgRNA_index, sgRNA_efficiency) %>% distinct %>% 
  arrange(sgRNA_index) %>% deframe
```


## Position bias of sgRNA repression

Plot the **weight of each sgRNA** to see if there is a dependency between correlation and sgRNA position. There is no significant trend.

```{r, fig.width = 5, fig.height = 4}
df_main %>%
  select(sgRNA_target, sgRNA_index, sgRNA_weight) %>%
  filter(sgRNA_index <= 5) %>%
  distinct %>%
  # plot
  ggplot(aes(x = factor(sgRNA_index), y = sgRNA_weight)) +
  geom_boxplot(outlier.shape = "") +
  stat_summary(fun.data = function(x) c(y = median(x)+0.07, 
    label = round(median(x), 2)), geom = "text") +
  stat_summary(fun.data = function(x) c(y = 1.2, 
    label = length(x)), geom = "text", color = grey(0.5)) +
  coord_cartesian(ylim = c(-0.25, 1.25)) +
  custom_theme()
```

We can also quantify how many genes have strongly correlated sgRNAs and how many have outliers. In order to do this, the median weight of the (up to) 5 sgRNAs per gene is plotted. Generally, the median weight ranges between 0.5 and 1.0, showing on average good correlation.

```{r, fig.width = 5, fig.height = 4}
df_main %>%
  select(sgRNA_target, sgRNA_index, sgRNA_weight) %>%
  filter(sgRNA_index <= 5) %>%
  distinct %>% group_by(sgRNA_target) %>%
  summarize(
    median_sgRNA_weight = median(sgRNA_weight),
    min_sgRNA_weight = min(sgRNA_weight)
  ) %>%
  # plot
  ggplot(aes(x = median_sgRNA_weight)) +
  geom_histogram(bins = 40, fill = custom_colors[1], alpha = 0.7) +
  custom_theme()
```

Second, the binding position of the sgRNAs could be correlated to the strength of repression. In other words sgRNAs binding closer to the promoter could have stronger ability to repress a gene, see Figure 1 B in [Wang et al., Nature Comms, 2018](https://www.nature.com/articles/s41467-018-04899-x). We plot **sgRNA efficiency** for genes only, because the absolute majority of those has 5 sgRNAs.

```{r, fig.width = 5, fig.height = 4}
df_main %>%
  filter(sgRNA_index <= 5, sgRNA_type == "gene") %>%
  select(sgRNA_target, sgRNA_index, sgRNA_efficiency) %>% distinct %>%
  ggplot(aes(x = factor(sgRNA_index), y = sgRNA_efficiency)) +
  geom_boxplot(notch = FALSE, outlier.shape = ".") +
  coord_cartesian(ylim = c(-0.25, 1.25)) +
  stat_summary(fun.data = function(x) c(y = median(x)+0.07, 
    label = round(median(x), 2)), geom = "text") +
  stat_summary(fun.data = function(x) c(y = 1.2, 
    label = length(x)), geom = "text", color = grey(0.5)) +
  custom_theme()
```

# Gene fitness calculation

## Summarize sgRNA fitness to gene fitness

With the correlation and the efficiency per sgRNA, we can compute the **weighted mean of all sgRNAs**. For comparison, we also test simple strategies such as the standard **arithmetic mean** and a top 1 and top 2 sgRNAs strategy. Metrics are calculated for log2 FC, and fitness.

```{r, warning = FALSE, message = FALSE}
df_gene <- df_main %>%
  
  # keep all annotation columns
  group_by(sgRNA_target, sgRNA_type, locus, gene_name, condition, time) %>%
  
  # summarize FC and fitness...
  summarize(.groups = "drop",
    # log2 FC
    mean_log2FoldChange = mean(log2FoldChange),
    wmean_log2FoldChange = weighted.mean(log2FoldChange, sgRNA_weight * sgRNA_efficiency),
    top1_log2FoldChange = log2FoldChange[which.max(sgRNA_efficiency)],
    top2_log2FoldChange = mean(log2FoldChange[order(sgRNA_efficiency, decreasing = TRUE)[1:2]]),
    sd_log2FoldChange = sd(log2FoldChange),
    # fitness
    mean_fitness = mean(fitness),
    wmean_fitness = weighted.mean(fitness, sgRNA_weight * sgRNA_efficiency),
    top1_fitness = fitness[which.max(sgRNA_efficiency)],
    top2_fitness = mean(fitness[order(sgRNA_efficiency, decreasing = TRUE)[1:2]]),
    sd_fitness = sd(fitness)
  )
```

A comparison of log2 FC aggregated by the different method shows clear differences. For the example gene `rps10` the weighted mean and the top method give similar results, representative of the stronger influence from highly depleted sgRNA repression strains. The regular mean is robust, but "shallow", probably underestimating the real effect n fitness. The top 1 method simply picks the most depleted/enriched sgRNA (over all conditions) as representative.

```{r, fig.width = 7, fig.height = 4}
df_gene %>% filter(sgRNA_target == "rps10") %>%
  pivot_longer(cols = matches("[n12]_log2FoldChange"), names_to = "metric", values_to = "log2FoldChange") %>%
  mutate(metric = str_remove(metric, "_log2FoldChange")) %>%
  ggplot(aes(x = time, y = log2FoldChange, 
    ymin = log2FoldChange-sd_log2FoldChange, 
    ymax = log2FoldChange+sd_log2FoldChange, color = fct_inorder(metric))) +
  geom_line(size = 1) + geom_point(size = 2) + geom_linerange(size = 1) +
  facet_wrap(~ condition, ncol = 3) +
  custom_theme(legend.position = "right") +
  coord_cartesian(ylim = c(-3.75, 0.75)) +
  scale_color_manual(values = custom_colors)
```

This plot shows a comparison of the 4 methods for the first 36 genes by alphabetical order, for one selected condition only (1% CO2, BG11, 1,000 µmol photons m-1 s-1). Here we can see that the top1 method is often but not always representative for the gene: For apcD or apcF, it does not seem representative compared to the mean, weighted mean, and top2 methods.

```{r, fig.width = 9, fig.height = 9}
df_gene %>% filter(
    gene_name %in% unique(.data[["gene_name"]])[1:36],
    condition == "1% CO2, 1000uE"
  ) %>%
  pivot_longer(cols = matches("[n12]_log2FoldChange"), names_to = "metric", values_to = "log2FoldChange") %>%
  mutate(metric = str_remove(metric, "_log2FoldChange")) %>%
  ggplot(aes(x = time, y = log2FoldChange, 
    ymin = log2FoldChange-sd_log2FoldChange,
    ymax = log2FoldChange+sd_log2FoldChange, color = fct_inorder(metric))) +
  geom_line(size = 1) + geom_point(size = 2) + geom_linerange(size = 1) +
  facet_wrap(~ sgRNA_target, ncol = 6) +
  custom_theme(legend.position = "top") +
  coord_cartesian(ylim = c(-5, 5)) +
  scale_color_manual(values = custom_colors)
```

The same control can be performed for gene fitness, that is, the AUC of log2 FC over all time points, for each individual sgRNA. Fitness was also summarized using the mean, weighted mean, or a top1 and top2 method for each of the 36 first genes by alphabetical order and one selected condition only. We can conclude from the plot below that mean and weighted mean only result in marginal difference in estimation of fitness, and continue with the weighted versions for all following analyses.

```{r, fig.width = 9, fig.height = 4.5}
df_gene %>% filter(
    gene_name %in% unique(.data[["gene_name"]])[1:36],
    condition == "1% CO2, 1000uE"
  ) %>%
  pivot_longer(cols = matches("[n12]_fitness"), names_to = "metric", values_to = "fitness") %>%
  mutate(metric = str_remove(metric, "_fitness")) %>% distinct %>%
  ggplot(aes(x = sgRNA_target, y = fitness, 
    ymin = fitness-sd_fitness, 
    ymax = fitness+sd_fitness, fill = fct_inorder(metric), color = fct_inorder(metric))) +
  geom_col(position = "dodge") +
  geom_errorbar(position = "dodge", size = 1) +
  custom_theme(legend.position = "top") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = custom_colors[1:4]) +
  scale_color_manual(values = custom_colors[1:4])
```

## Global distribution of gene fitness

Global distribution of weighted mean fitness for all genes. Effect of ncRNA repression seems to be much lower than effect of gene repression.

```{r, fig.width = 7, fig.height = 4}
df_gene %>% filter(time == 0) %>%
  ggplot(aes(x = wmean_fitness, fill = sgRNA_type)) +
  geom_histogram(bins = 100) +
  coord_cartesian(xlim = c(-2, 2), ylim = c(0, 1000)) +
  facet_wrap( ~ condition, ncol = 3) +
  custom_theme() +
  scale_fill_manual(values = custom_colors[3:4])
```


## Behavior of control sgRNAs

Ten sgRNAs were included in the library that have no gene-specific/random (?) targets. The following plot shows that these negative controls do not have an effect on strain fitness, except probably 2 sgRNAs in one specific condition.

```{r, fig.width = 7, fig.height = 4}
df_main %>% filter(grepl("ctrl", sgRNA_target)) %>%
  ggplot(aes(x = time, y = log2FoldChange, color = sgRNA_target)) +
  geom_line(size = 1) + geom_point(size = 2) + ylim(-5, 5) +
  facet_wrap(~ condition, ncol = 3) +
  custom_theme() +
  scale_color_manual(values = rep_along(1:10, custom_colors))
```

# Unsupervised clustering of genes

## Cluster genes by similarity

```{r, fig.width = 8, fig.height = 5}
# generate color palette for heatmap
heat_cols <- colorspace::diverging_hcl(n = 7, h = c(255, 12), c = c(50, 80), l = c(20, 97), power = c(1, 1.3))

# create a matrix-like df with wide fitness data for plotting heatmap
df_heatmap <- df_gene %>%
  filter(time == 0, !is.na(locus)) %>%
  select(locus, condition, wmean_fitness) %>%
  mutate(wmean_fitness = wmean_fitness %>% replace(., . > 4, 4) %>% replace(., . < -4, -4)) %>%
  pivot_wider(names_from = condition, values_from = wmean_fitness) %>%
  column_to_rownames(var = "locus")

# subset of df with *significantly changed* genes
df_heatmap2 <- df_heatmap %>%
  filter(if_any(.cols = matches("uE$"), ~ !between(., -2, 2)))

# create cluster for reordering
mat_cluster <- df_heatmap %>% as.matrix %>% dist %>% hclust(method = "ward.D2")
mat_heatmap <- df_heatmap %>% as.matrix %>%
  .[order.dendrogram(as.dendrogram(mat_cluster)), ncol(.):1]

# repeat this step with subset of *significantly changed* genes
mat_cluster_sig <- df_heatmap2 %>% as.matrix %>% dist %>% hclust(method = "ward.D2")
mat_heatmap_sig <- df_heatmap2 %>% as.matrix %>%
  .[order.dendrogram(as.dendrogram(mat_cluster_sig)), ncol(.):1]
```


Now we can plot _all_ genes, a subset with only significant genes, and a dendrogram for clustering.
 
```{r, fig.width = 8, fig.height = 1.6}
levelplot(mat_heatmap,
  par.settings = custom.colorblind(),
  col.regions = colorRampPalette(heat_cols)(20),
  at = seq(-4, 4, 0.5), aspect = "fill",
  xlab = paste0("sgRNA (", nrow(mat_heatmap),")"),
  ylab = "", scales = list(x = list(draw = FALSE)),
  panel = function(x, y, z, ...) {
    panel.levelplot(x, y, z, ...)
    panel.abline(h = 1:5+0.5, col = "white", lwd = 1.5)
  }
)
```

```{r, fig.width = 8, fig.height = 3}
plot_heatmap_sig <- levelplot(mat_heatmap_sig,
  par.settings = custom.colorblind(),
  col.regions = colorRampPalette(heat_cols)(20),
  at = seq(-4, 4, 0.5), aspect = "fill",
  xlab = paste0("sgRNA (", nrow(mat_heatmap_sig),")"),
  ylab = "", scales = list(x = list(draw = FALSE)),
  panel = function(x, y, z, ...) {
    panel.levelplot(x, y, z, ...)
    panel.abline(h = 1:5+0.5, col = "white", lwd = 1.5)
  }
)

plot_cluster_dend <- mat_cluster_sig %>% as.dendrogram %>%
  set("branches_k_col", custom_colors[1:5], k = 5) %>%
  set("branches_lwd", 0.5) %>%
  as.ggdend %>%
  ggplot(labels = FALSE)

gridExtra::grid.arrange(
  # coords for unit: top, right, bottom, left
  plot_cluster_dend + 
    theme(plot.margin = unit(c(0.1, 0.08, -0.35, 0.28),"npc")),
  plot_heatmap_sig,
  nrow = 2
)
```

Clustering shows that the depletion/enrichment patterns in different conditions are quite similar. Different clusters entail genes with:

- genes heavily depleted in all conditions
- genes somewhat depleted in all conditions
- genes depleted **primarily in low light**
- genes depleted **primarily in low light + glucose (+DCMU)**
- genes enriched **in high CO2, high light**

```{r}
df_gene <- mat_cluster_sig %>%
  cutreeord(k = 5) %>% enframe("locus", "cluster") %>%
  mutate(cluster_order = match(locus, mat_cluster_sig$labels[mat_cluster_sig$order])) %>%
  right_join(df_gene, by = "locus")

df_gene %>% select(locus, cluster) %>%
  distinct %>% count(cluster)
```

## Visualize gene similarity by dimensionality reduction methods

We use two different dimensionality reduction methods, **nMDS** and **t-SNE**. We can check if these methods reproduce the clustering for the significantly regulated genes produced with `hclust`. Analysis shows that the small clusters are more strongly separated from the rest.

```{r, fig.width = 8, fig.height = 4}
# set a seed to obtain same pattern for stochastic methods
set.seed(123)

# run nMDS analysis
NMDS <- df_heatmap2 %>% as.matrix %>% dist %>% metaMDS
df_nmds <- NMDS$points %>% as_tibble(rownames = "locus") %>%
  left_join(df_gene %>% select(locus, cluster) %>% distinct)

# run t-SNE analysis
SNE <- tsne::tsne(df_heatmap2 %>% as.matrix %>% dist)
df_tsne <- SNE %>% setNames(c("x", "y")) %>% as_tibble %>%
  mutate(locus = rownames(df_heatmap2)) %>%
  left_join(df_gene %>% select(locus, cluster) %>% distinct)

plot_nmds <- df_nmds %>% 
  ggplot(aes(x = MDS1, y = MDS2, color = factor(cluster))) +
  geom_point(size = 2) + labs(title = "nMDS") +
  custom_theme(legend.position = c(0.85, 0.25)) +
  scale_color_manual(values = custom_colors)

plot_tsne <- df_tsne %>% 
  ggplot(aes(x = V1, y = V2, color = factor(cluster))) +
  geom_point(size = 2) + labs(title = "t-SNE") +
  custom_theme(legend.position = c(0.85, 0.78)) +
  scale_color_manual(values = custom_colors)

gridExtra::grid.arrange(plot_nmds, plot_tsne, ncol = 2)
```


## Extract and analyze most interesting gene clusters

**Cluster 1**: Genes strongly enriched **during high light exposure** (1% CO2, 1000 µmol photons m^-2 s^-1)

```{r, fig.width = 7.5, fig.height = 3.5}
df_gene %>% filter(cluster == 1, time == 0) %>%
  ggplot(aes(x = sgRNA_target, y = wmean_fitness, 
    ymin = wmean_fitness-sd_fitness, 
    ymax = wmean_fitness+sd_fitness, fill = condition, color = condition)) +
  geom_col(position = "dodge") +
  geom_errorbar(position = "dodge", size = 1) +
  custom_theme() + labs(title = "Mutants enriched in high light/CO2") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  scale_fill_manual(values = custom_colors) +
  scale_color_manual(values = custom_colors)
```

More space-efficient representation as a heat map.

```{r, fig.width = 5.0, fig.height = 2.5}
df_gene %>% filter(cluster == 1, time == 0) %>%
  arrange(cluster_order) %>% 
  mutate(sgRNA_target = factor(sgRNA_target, unique(sgRNA_target))) %>%
  mutate(wmean_fitness = wmean_fitness %>% replace(., . > 3, 3) %>% replace(., . < -3, -3)) %>%
  ggplot(aes(x = sgRNA_target, y = condition, fill = wmean_fitness)) +
  geom_tile() + custom_theme() +
  labs(title = "High light/CO2", x = "", y = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  scale_fill_distiller(palette = "RdBu", limits = c(-3, 3))
  #palettes: BrBG, PiYG, PRGn, PuOr, RdBu, RdGy, RdYlBu, RdYlGn, Spectral
```

**Cluster 5** Genes depleted **on glucose + low light** (0.05 or 1% CO2, glucose (+DCMU), 60 µmol photons m^-2 s^-1)

```{r, fig.width = 10.5, fig.height = 2.5}
df_gene %>% 
  filter(cluster == 5, time == 0) %>%
  arrange(cluster_order) %>% 
  mutate(sgRNA_target = factor(sgRNA_target, unique(sgRNA_target))) %>%
  mutate(wmean_fitness = wmean_fitness %>% replace(., . > 3, 3) %>% replace(., . < -3, -3)) %>%
  ggplot(aes(x = sgRNA_target, y = condition, fill = wmean_fitness)) +
  geom_tile() + custom_theme() +
  labs(title = "Mutants depleted on glucose + low light", x = "", y = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  scale_fill_distiller(palette = "RdBu", limits = c(-3, 3))
```

**Cluster 4**: Genes depleted **on low light** (0.05 or 1% CO2, 60 µmol photons m^-2 s^-1)

```{r, fig.width = 8.5, fig.height = 2.5}
df_gene %>% filter(cluster == 4, time == 0) %>%
  arrange(cluster_order) %>% 
  mutate(sgRNA_target = factor(sgRNA_target, unique(sgRNA_target))) %>%
  mutate(wmean_fitness = wmean_fitness %>% replace(., . > 3, 3) %>% replace(., . < -3, -3)) %>%
  ggplot(aes(x = sgRNA_target, y = condition, fill = wmean_fitness)) +
  geom_tile() + custom_theme() +
  labs(title = "Mutants depleted on low light", x = "", y = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  scale_fill_distiller(palette = "RdBu", limits = c(-3, 3))
```

- **Apc and cpc repression mutants** encoding phycobilisomes are also enriched in high light


```{r, fig.width = 6.5, fig.height = 2.5}
df_gene %>% filter(str_detect(gene_name, "[ac]pc"), time == 0) %>%
  mutate(wmean_fitness = wmean_fitness %>% replace(., . > 3, 3) %>% replace(., . < -3, -3)) %>%
  ggplot(aes(x = sgRNA_target, y = condition, fill = wmean_fitness)) +
  geom_tile() + custom_theme() +
  labs(title = "Apc/Cpc mutants enriched in high light/CO2", x = "", y = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  scale_fill_distiller(palette = "RdBu", limits = c(-3, 3))
```

## Interaction analysis using STRING database

The next step is to retrieve additional information about these clusters to find correlated genes, groups, pathways, or regulatory networks. We can use the STRING database for this task.

```{r, echo = FALSE}
suppressPackageStartupMessages({
  library(ggraph)
  library(tidygraph)
})

# function to retrieve network interaction data from STRING DB
# separate gene IDs by "%0d"; species/taxon ID for Synechocystis sp. PCC 6803: 1148
# (see https://string-db.org/cgi/organisms)
plot_STRING <- function(gene_ID, taxon_ID, 
  min_score = 0000, ref = NULL, plot = TRUE, groups = NULL
) {
  gene_list <- paste(gene_ID, collapse = "%0d")
  string_graph <- paste0(
    "https://string-db.org/api/tsv/network?identifiers=", 
    gene_list, "&species=", taxon_ID, "&required_score=", min_score) %>%
    read_tsv(col_types = cols()) %>%
    select(-stringId_A, -stringId_B) %>%
    as_tbl_graph()
  if (!is.null(ref)) {
    # add gene annotation
    string_graph <- left_join(string_graph, ref, by = "name")
  }
  if (plot) {
    # function to space labels in certain distance to circle
    nudge_circle <- function(n, size_x = 0.2, size_y = 0.15) {
      nudge_x = size_x * -cos(seq(0.5*pi, 2.5*pi, length.out = n))
      nudge_y = size_y * sin(seq(0.5*pi, 2.5*pi, length.out = n))
      list(x = nudge_x, y = nudge_y)
    }
    # plot using ggraph
    n_nodes <- string_graph %>% activate(nodes) %>% pull(name) %>% length
    string_graph %>% arrange(across(matches(groups))) %>% activate(edges) %>%
      filter(score >= 0.4) %>%
      ggraph(layout = 'linear', circular = TRUE) +
      geom_edge_arc(colour = grey(0.6, 0.5), aes(width = score)) + 
      geom_node_point(aes(colour = .data[[groups]]), size = 4) +
      geom_node_text(nudge_x = nudge_circle(n_nodes)$x, 
        nudge_y = nudge_circle(n_nodes)$y,
        size = 3, aes(label = name, colour = .data[[groups]])) +
      scale_edge_width(range = c(0.2, 2)) +
      theme_graph(background = "white", foreground = grey(0.5),
        plot_margin = margin(10, 10, 10, 10))
  }
}
```


```{r, fig.width = 5.5, fig.height = 4.0}
plot_STRING(
  gene_ID = filter(df_gene, cluster == 4) %>% pull(locus) %>% unique,
  taxon_ID = 1148,
  ref = df_gene %>% select(sgRNA_target, locus, gene_name) %>% 
    distinct %>% rename(name = sgRNA_target),
  groups = "name"
)
```


```{r, fig.width = 6.8, fig.height = 4.7}
plot_STRING(
  gene_ID = filter(df_gene, cluster == 5) %>% pull(locus) %>% unique,
  taxon_ID = 1148,
  ref = df_gene %>% select(sgRNA_target, locus, gene_name) %>% 
    distinct %>% rename(name = sgRNA_target),
  groups = "locus"
)
```

# Direct comparison of gene fitness

## Fitness of all conditions vs each other

The scatterplot matrix shows that correlation between conditions is high; this is expected as a similar set of genes is essential in most conditions. It also shows that one condition, `0.05% CO2, 200 µE` sticks out as more shallow, that means on average lower response in fitness. This is most likely an artifact of differing number of generations leading to slower depletion compared to other samples.

```{r, fig.width = 8, fig.height = 8}
custom_splom(df_heatmap,
  prepanel.limits = function(x) {c(-4, 1)},
  col_palette = rainbow(4),
  ylim = c(-4, 1), xlim = c(-4, 1),
  xlab = "data points", ylab = "regression",
  pch = 19, col = grey(0, 0.3), cex = 0.2
)
```


----------

We can plot selected conditions against each other and add gene labels in order to find patterns.

```{r, fig.width = 8, fig.height = 12}
make_fitness_plot <- function(data, vars, title = NULL) {
  data %>% ungroup %>%
    filter(condition %in% vars, sgRNA_type == "gene") %>%
    select(locus, sgRNA_target, condition, wmean_fitness) %>% distinct %>%
    group_by(sgRNA_target) %>% 
    mutate(significant = abs(wmean_fitness[1]-wmean_fitness[2])>2) %>%
    pivot_wider(names_from = condition, values_from = wmean_fitness) %>%
    mutate(sgRNA_target = if_else(significant, sgRNA_target, "")) %>%
    # ggplot
    ggplot(aes(x = get(vars[1]), y = get(vars[2]), 
      color = significant, label = sgRNA_target)) +
    geom_point(size = 1) + custom_theme(legend.position = 0) +
    geom_abline(intercept = 0, slope = 1, col = grey(0.5), lty = 2, size = 0.8) +
    geom_abline(intercept = 2, slope = 1, col = grey(0.5), lty = 2, size = 0.8) +
    geom_abline(intercept = -2, slope = 1, col = grey(0.5), lty = 2, size = 0.8) +
    geom_text_repel(size = 3, max.overlaps = 50) +
    labs(title = title, x = vars[1], y = vars[2]) +
    coord_cartesian(xlim = c(-4.5, 2.5), ylim = c(-4.5, 2.5)) +
    scale_color_manual(values = c(grey(0.5), custom_colors[2]))
}

# CONDITIONS:
# "0.05% CO2, 60uE"   "0.05% CO2, dcmu, glucose, 60uE"
# "1% CO2, 1000uE"     "1% CO2, 60uE"      "1% CO2, Glucose, 60uE"
gridExtra::grid.arrange(nrow = 3, ncol = 2,
  make_fitness_plot(df_gene, c("0.05% CO2, 60uE", "0.05% CO2, dcmu, glucose, 60uE"),
    title = "low CO2, photo-heterotrophy"),
  make_fitness_plot(df_gene, c("1% CO2, 60uE", "0.05% CO2, dcmu, glucose, 60uE"),
    title = "high CO2, photo-heterotrophy"),
  make_fitness_plot(df_gene, c("1% CO2, 60uE", "1% CO2, 1000uE"),
    title = "high CO2, low vs high light"),
  make_fitness_plot(df_gene, c("1% CO2, 60uE", "1% CO2, Glucose, 60uE"),
    title = "high CO2, mixotrophy"),
  make_fitness_plot(df_gene, c("1% CO2, 60uE", "0.05% CO2, 60uE"),
    title = "low light, low vs high CO2")
)
```

# Differential fitness of selected gene sets

## Central carbon metabolism

To plot gene fitness for the enzymes of central carbon metabolism, we need a complete list of enzymes and the genes that they are mapped to. To list the different **KEGG databases** that can be queried, use `listDatabases()`. Gene-pathway mappings are obtained and merged with pathway names and gene/enzyme names.

```{r}
# get mapping of pathways for each gene
df_kegg <- keggLink("pathway", "syn") %>%
  enframe(name = "locus", value = "kegg_pathway_id") %>%
  
  # get list of pathways with name/ID pairs
  left_join(by = "kegg_pathway_id",
    keggList("pathway", "syn") %>%
    enframe(name = "kegg_pathway_id", value = "kegg_pathway")
  ) %>%
  
  # get list of gene/enzyme names
  left_join(by = "locus",
    keggList("syn") %>%
    enframe(name = "locus", value = "kegg_gene") %>%
    mutate(kegg_gene_short = str_extract(kegg_gene, "^[a-zA-Z0-9]*;") %>% 
      str_remove(";"))
  ) %>%
  
  # trim useless prefixes
  mutate(
    locus = str_remove(locus, "syn:"),
    kegg_pathway_id = str_remove(kegg_pathway_id, "path:"),
    kegg_pathway = str_remove(kegg_pathway, " - Synechocystis sp. PCC 6803")
  )

head(df_kegg)
```

Then we can extract gene sets for specific pathways and plot fitness. We start with glycolysis and Calvin cycle enzymes.

```{r}
list_central_met_pathways <- c(
  "Glycolysis / Gluconeogenesis",
  "Pentose phosphate pathway",
  "Carbon fixation in photosynthetic organisms",
  "Photosynthesis",
  "Photosynthesis - antenna proteins",
  "Citrate cycle (TCA cycle)",
  "Pyruvate metabolism",
  "Glyoxylate and dicarboxylate metabolism"
)
```


```{r}
plot_gene_fitness <- lapply(list_central_met_pathways, function(pw) {
  df_gene %>% filter(time == 0) %>%
    inner_join(df_kegg %>% filter(kegg_pathway == pw) %>% select(locus),
      by = "locus") %>%
    
    ggplot(aes(x = condition, y = wmean_fitness, 
      ymin = wmean_fitness-sd_fitness, 
      ymax = wmean_fitness+sd_fitness, fill = condition, color = condition)) +
      geom_col(position = "dodge", width = 0.6) +
      geom_errorbar(position = "dodge", width = 0.6, size = 1) +
      custom_theme(aspect.ratio = 1,
        legend.position = "bottom", legend.key.size = unit(0.4, "cm")) + 
      labs(title = pw) +
      theme(axis.text.x = element_blank()) +
      scale_fill_manual(values = custom_colors) +
      scale_color_manual(values = custom_colors) +
      facet_wrap(~ sgRNA_target, ncol = 8)
})
```


```{r, fig.width = 8, fig.height = 6}
print(plot_gene_fitness[[1]])
```

```{r, fig.width = 8, fig.height = 5}
print(plot_gene_fitness[[2]])
```

```{r, fig.width = 8, fig.height = 5}
print(plot_gene_fitness[[3]])
```

```{r, fig.width = 8, fig.height = 11}
print(plot_gene_fitness[[4]])
```

```{r, fig.width = 8, fig.height = 4}
print(plot_gene_fitness[[5]])
```

```{r, fig.width = 8, fig.height = 4}
print(plot_gene_fitness[[6]])
```

**Custom set of stress related genes**

OCP (`slr1963`), pgr5 (`ssr2016`), Flv1 (`sll1521`), Flv2 (`sll0219`), Flv3 (`sll0550`), Flv4 (`sll0217`), `sll0218` (in flv2/4 operon), SigB (`sll0306`), SigC (`sll0184`), SigD (`sll2012`), SigE (`sll1689`).

```{r, echo = FALSE}
plot_stress_genes <- df_gene %>%
  filter(time == 0, locus %in% c("slr1963", "ssr2016", "sll1521",
    "sll0219", "sll0550", "sll0217", "sll0218", "sll0306", "sll0184",
    "sll2012", "sll1689")) %>%
  
  ggplot(aes(x = condition, y = wmean_fitness, 
    ymin = wmean_fitness-sd_fitness, 
    ymax = wmean_fitness+sd_fitness,
    fill = condition, color = condition)) +
    geom_col(position = "dodge", width = 0.6) +
    geom_errorbar(position = "dodge", width = 0.6, size = 1) +
    custom_theme(aspect.ratio = 1,
      legend.position = "bottom", legend.key.size = unit(0.4, "cm")) + 
    labs(title = "Stress related genes") +
    theme(axis.text.x = element_blank()) +
    scale_fill_manual(values = custom_colors) +
    scale_color_manual(values = custom_colors) +
    facet_wrap(~ sgRNA_target, ncol = 8)
```


```{r, fig.width = 8, fig.height = 4}
print(plot_stress_genes)
```

```{r, include = FALSE, eval = FALSE, fig.width = 7, fig.height = 4}
df_main %>% filter(grepl("gltA", sgRNA_target)) %>%
  ggplot(aes(x = time, y = log2FoldChange, color = sgRNA)) +
  geom_line(size = 1) + geom_point(size = 2) + ylim(-5, 5) +
  facet_wrap(~ condition, ncol = 3) +
  custom_theme() +
  scale_color_manual(values = rep_along(1:10, custom_colors))
```


# Gene enrichment analysis

Sometimes even small effects in fitness can be relevant if several genes of the same pathway (or iso-enzymes) are affected. A simple fitness threshold will not reveal those changes. In such cases a more nuanced approach can be taken, a gene set enrichment analysis (GSEA). Several packages exist to test if functionally related genes are enriched, depleted, or both at the same time / the same conditions.

Before we test for enrichemnt of associated pathways/GO terms, we can have a look at the general depletion/enrichment per KEGG pathway. The fitness distribution per pathway can be visualized using a violin- or scatter plot.

```{r, fig.width = 10, fig.height = 15}
df_gene %>% filter(time == 0) %>%
  inner_join(df_kegg, by = "locus") %>%
  group_by(kegg_pathway, condition) %>%
  filter(length(unique(locus)) > 10) %>%
  
  ggplot(aes(x = condition, y = wmean_fitness,
      fill = condition, color = condition)) +
    geom_abline(intercept = 0, slope = 0, lty = 2, color = grey(0.5)) +
    geom_violin(color = "white") +
    stat_summary(fun.data = mean_sdl, geom = "pointrange", size = 0.4) +
    custom_theme(aspect.ratio = 1,
      legend.position = "bottom", 
      legend.key.size = unit(0.4, "cm")) +
    theme(axis.text.x = element_blank()) +
    facet_wrap(~ str_sub(kegg_pathway, 1, 16), ncol = 7) +
    coord_cartesian(ylim = c(-3.5, 3.5)) +
    scale_fill_manual(values = alpha(custom_colors, 0.5)) +
    scale_color_manual(values = custom_colors)
```

## KEGG pathway enrichment

We use the functions `kegga` for KEGG enrichemnt analysis and `goana` for GO term enrichment from the `limma` package. Both functions test for over or under-representation of genes associated with certain pathways or GO terms. The functions don't take the strength of differential fitness into account (DF; the depletion/enrichment over time).


```{r}
df_kegg_enrichment <- lapply(unique(df_gene$condition), function(cond) {
  df_gene %>% filter(
  sgRNA_type == "gene", time == 0,
  condition == cond) %>%
  
  # filter for differential fitness (DF) genes
  filter(!between(wmean_fitness, -1.0, 1.0), !is.na(locus)) %>%
  
  # perform KEGG enrichment
  pull(locus) %>% kegga(species.KEGG = "syn") %>%
  mutate(condition = cond)
}) %>% bind_rows

head(df_kegg_enrichment)
```

Now we visualize the pathways that are most enriched for DF genes. It turns out that ribosomal proteins are extremely depleted and therefore score high on the negative log10 p-value for pathway enrichment.

```{r}
df_kegg_enrichment %>%
  group_by(Pathway) %>% filter(any(P.DE <= 0.01)) %>%
  select(Pathway, condition, P.DE) %>%
  mutate(log10_p_value = -log10(P.DE), .keep = "unused") %>%
  
  # format some columns
  mutate(condition = str_replace(condition, pattern = "0.05% CO2", replacement = "LC") %>%
    str_replace(pattern = "1% CO2", replacement = "HC") %>%
    str_replace(pattern = "60uE", replacement = "LL") %>%
    str_replace(pattern = "1000uE", replacement = "HL")) %>%
  mutate(Pathway = str_sub(Pathway, 1, 25)) %>%
  # optionally remove ribosomes
  # filter(Pathway != "Ribosome") %>%
  
  # make correlation plot
  pivot_wider(names_from = condition, values_from = log10_p_value) %>%
  column_to_rownames(var = "Pathway") %>% as.matrix %>% t %>%
  corrplot(is.corr = FALSE, tl.col = grey(0.5), tl.cex = 0.8,
    col = colorRampPalette(custom_colors[c(1,5,2)])(10), col.lim = c(0, 20))
```


# Export summary table of all genes and conditions

Export a summary table of all genes and conditions, so that it's easy for other people to look up single conditions as for example done in [one-by-one fitness comparisons](#fitness-of-all-conditions-vs-each-other). This is best done in wide format (one column per condition).

```{r}
df_gene %>% ungroup %>%
  filter(sgRNA_type == "gene") %>%
  select(locus, sgRNA_target, gene_name, condition, wmean_fitness) %>% 
  distinct %>%
  pivot_wider(names_from = condition, values_from = wmean_fitness) %>%
  write_csv("../data/output/fitness_summary.csv")

df_gene %>%
  filter(sgRNA_type == "gene") %>%
  write_csv("../data/output/fitness_genes.csv")
```


# Session Info

```{r}
sessionInfo()
```

