---
title: "Mapping KEGG pathways to proteins"
output: html_notebook
---

### Investigating protein allocation in _Ralstonia eutropha_ H16

The _Ralstonia eutropha_ H16 genome was previously annotated using the EMBL-maintained tool 
[eggNOG](http://eggnogdb.embl.de/#/app/home), see my R notebook 
```../Ralstonia_COG/Ralstonia_H16_genome_re_annotation.Rmd ```. 
All genes/proteins were annotated based on sequence similarity and grouped 
into homology groups (Cluster of Orthologous Groups, COG). Groups were then
manually curated by moving genes from less to better matching categories. 

However, one important difference between COG and KEGG is that KEGG Brite annotation 
is not unambiguous like COG, but one gene/protein can be mapped to multiple 
functions/pathways. This behavior is more realistic then the COG classification.
Another advantage is that KEGG is a manually curated, systematic, and well-connected
database. Every protein in KEGG is associuated with enzyme-EC numbers, GO terms, 
pathways, and associated enzymatic reactions. This makes KEGG pathways
superior to COG terms when it comes to funtional interpretation or quantitative
metabolic models.

In this notebook I will try to map KEGG pathways (Brite or Modules) to the _R. eutropha_
proteome and deal with ambiguous protein annotation. Sources that I use are:

- Previously expanded _R. eutropha_ annotation based on COG ([eggNOG](http://eggnogdb.embl.de/#/app/home), downloaded 2019-06-12)
- [KEGG BRite](https://www.kegg.jp/kegg-bin/show_organism?org=reh) database for _R. eutropha_ H16 (downloaded 2018-09-10)

### Problem definition 

1. Map proteins to KEGG pathways, either 'Brite' hierarchy or 'Modules'

2. Investigate predicted protein abundance range per pathway (min, mean, max)

3. Try to formalize biological role of pathways: size, reactions, input, output, variability


### Task 1: Map proteins to KEGG pathways

We begin with elucidating differences between KEGG Brite and KEGG Modules. Are they the same 
(apparently not), which one is more comprehensive or has higher coverage for _R. eutropha_?


```{r, message = FALSE}
# load packages
library(tidyverse)
library(lattice)
library(latticeExtra)
library(Rtools)
library(stringr)
library(graph)
library(igraph)
library(KEGGREST)
library(KEGGgraph)
```

Several alternatives are available to download and/or parse KEGG xml files,
among them the R `KEGGREST` or the `KEGGgraph` package. One can also use simple bash
download from KEGG using their REST API (thanks Johannes Asplund-Samuelsson for 
simple example). 

```{bash, eval = FALSE}
wget -qO - http://rest.kegg.jp/link/ko/ko01200
```

We can also import KEGG files directly into R data frames or lists
using the KEGGREST package which provides an R interface for the KEGG API.

```{r, messagev= FALSE}
# KEGG Brite annotation file, manually parsed from website
reh_brite <- read_csv("../data/KEGG_Brite_20180910_reh00001.csv")
head(reh_brite)

# Alternatives
# ---------------
#
# KEGGREST package as interface to KEGG
# download annotation for all Ralstonia eutropha genes

# examples to query a single gene ID, KO number, module, or pathway
#keggGet("reh:H16_A0003") %>% head
#keggGet("K00847") %>% {.[[1]][1:5]} %>% head
#keggGet("M00432") %>% flatten
#keggGet("reh00010") %>% head

# obtain mapping of modules for each gene
reh_modules <- keggLink("module", "reh") %>% 
  enframe(name = "gene", value = "module")

# obtain mapping of modules for each gene
reh_pathways <- keggLink("pathway", "reh") %>% 
  enframe(name = "gene", value = "pathway")

# How many unique genes or modules in R.e.?
paste(
  reh_modules %>% pull(gene) %>% unique %>% length, "genes annotated in",
  reh_modules %>% pull(module) %>% unique %>% length, "modules"
)

# How many unique genes or pathways in R.e.?
paste(
  reh_pathways %>% pull(gene) %>% unique %>% length, "genes annotated in",
  reh_pathways %>% pull(pathway) %>% unique %>% length, "pathways"
)
  
# compared to KEGG Brite, downloaded manually
paste(
  reh_brite %>% pull(Protein) %>% unique %>% length, "genes annotated in",
  reh_brite %>% pull(Pathway) %>% unique %>% length, "KEGG Brite pathways"
)
```

----------

### Mapping genes expression to KEGG pathways

KEGG Modules are computationally generated units of KEGG pathways, while the latter
are manually constructed. For now, the KEGG DB contains many more gene mappings for pathways
then modules, which makes pathways the more promising and more complete base for
construction of a coarse-grained model. We continue with exploration of KEGG pathways
by adding Pathway names and meta data to the data frame. The final goal is to map gene expression quantifications to pathways and explore how many resources are invested per pathway,
and how resource allocation changes over conditions. First, some pathways are removed as they
are only overview 'maps'.

```{r}
# some pathways are overview pathways that don't add any information
# (aggregates of several pathways):
# 01100 - Metabolic pathways
# 01110 - Biosynthesis of secondary metabolites
# 01120 - Microbial metabolism in diverse environments
# 01130 - Biosynthesis of antibiotics
# 01200 - Carbon metabolism
# 01210 - 2-Oxocarboxylic acid metabolism
# 01212 - Fatty acid metabolism
# 01230 - Biosynthesis of amino acids
# 01220 - Degradation of aromatic compounds
reh_pathways <- reh_pathways %>% 
  filter(!grepl("01(100|110|120|130|200|210|212|220|230)", pathway))

# check how many unique genes remain in data
reh_pathways %>% pull(gene) %>% unique %>% length
```

### Pathway network topology

To retrieve the topology of the network, we can use the KEGGgraph package.
First we download all KGML files for all pathways. These files contain all 
nodes (enzymes) and edges (reactions/metabolites) of a metabolic network.

```{r, eval = FALSE}
# download KGML files for all pathways
for (pw in unique(reh_pathways$pathway)) {
  pw <- gsub("path:", "", pw)
  retrieveKGML(pw, organism = "reh", destfile = paste0("../KGML/", pw, ".kgml"))
}
```

The next step is to parse the graphs into an R graph or data frame object. Of these
two different possibilities, a graph is probably the better option as it allows to 
use other graph based functions. We consctruct two helper functions, one that parses
KGML files using KEGGgraph functions and stores results in a list; And a second 
that extracts the names of adjacent pathways that a pathway is connected to, 
in order to make a new coarse-grained, simplified network of all pathways.
A third function retrieves KEGG pathway annotation using KEGGREST and
returns a list.

```{r, fig.width = 12, fig.height = 10}
# function to parse all graphs from KGML files and store result in list
parse_KGML_list <- function(file_list) {
  graphs <- lapply(
    file_list, 
    function(x) {
      parseKGML2Graph(
        file = x,
        expandGenes = TRUE,
        genesOnly = FALSE
      )
    }
  )
  names(graphs) <- paste0(
    "path:",
    stringr::str_extract(
      file_list, "[a-z]{3}[0-9]*"
    )
  )
  graphs
}

# function to retrieve all other pathways that a pathway is connected to.
# the only required argument is a list of pathways represented
# as graph (actually graphNEL) object
get_adjacent_pathways <- function(pathway_graphs) {
  lapply(1:length(pathway_graphs),
    function(i) {
      pws <- grep("path:", nodes(pathway_graphs[[i]]), value = TRUE)
      data.frame(
        stringsAsFactors = FALSE,
        from = rep(names(pathway_graphs)[[i]], length(pws)),
        to = pws
      ) %>% filter(
        !to == from,
        !grepl(":map", to),
        !grepl(":map", from)
      )
    }
  ) %>% bind_rows
}

# function to retrieve pathway metadata from KEGG
# for arbitrary pathways (by name) and store results
# in list
get_pathway_metadata <- function(pathway_list) {
  res <- lapply(pathway_list, function(x) {
    KEGGREST::keggGet(x)[[1]]
  })
  names(res) <- pathway_list
  res
}
```

----------

The next step is to obtain a test set of pathways, such as the first 10 pathways
of _Ralstonia eutropha_, store them in a list and query all adjacent pathways.

```{r}
# parse pathways as graphs and store in list
reh_pathways_as_graph <- parse_KGML_list(
  list.files("../KGML", full.names = TRUE)
)

# extract all pathway to pathway connections
reh_pathways_connections <- get_adjacent_pathways(reh_pathways_as_graph)
head(reh_pathways_connections)

# get pathway meta data for all unique nodes
reh_pathways_nodes <- c(reh_pathways_connections) %>% unlist %>% unname %>% unique
reh_pathway_meta <- get_pathway_metadata(reh_pathways_nodes)
```

We can construct a new graph from the list of edges/connections and plot the 
network using igraph/tidygraph packages.
The input are two lists or data frames, 'nodes' with unique name and ID of genes
and 'edges' indicating all recorded connections between nodes; However,
the information in 'edges' already contains all possible nodes, so 'edges' alone can be sufficient. Node metadata can be stored in an extra data.frame (slot of the igraph object)

```{r, message = FALSE}
# load additional libraries
library(tidygraph)
library(ggplot2)
library(ggraph)

# make a data.frame with node metadata, handy for creating the graph
reh_pathways_nodes <- reh_pathways %>%
  filter(pathway %in% reh_pathways_nodes) %>%
  group_by(pathway) %>%
  summarize(number_of_genes = length(gene)) %>%
  left_join(.,
    reh_pathways_connections$from %>%
      table %>% as.list %>% unlist %>%
      enframe("pathway", "interactions")
  ) %>%
  left_join(.,
    lapply(reh_pathway_meta, function(x) {
      data.frame(
        stringsAsFactors = FALSE,
        pathway = paste0("path:", unname(x$ENTRY)),
        pathway_name = x$NAME,
        pathway_map = x$PATHWAY_MAP,
        pathway_map_short = gsub(" biosynthesis| metabolism", "", x$PATHWAY_MAP),
        pathway_class = x$CLASS
      ) 
    }) %>% bind_rows
  )

head(reh_pathways_nodes)

# construct a new graph object from of pathway nodes/edges
reh_pathways_igraph <- tbl_graph(
  edges = reh_pathways_connections,
  nodes = rename(reh_pathways_nodes, name = pathway)
)
```

Plot the graph.

```{r, fig.width = 10, fig.height = 8}
# the layout argument takes one of: 'nicely', 'star', 'circle', 'gem', 'dh', 
# 'randomly', 'fr', 'kk', 'drl', 'lgl', 'graphopt', 'grid', 'mds'
reh_pathways_plot <- ggraph(reh_pathways_igraph, layout = 'dh') +
geom_edge_link(colour = grey(0.5)) + 
geom_node_point(aes(colour = "orange", size = interactions)) +
geom_node_text(aes(label = pathway_map_short %>% gsub(" ", "\n", .)), repel = TRUE, col = grey(0.5), size = 2) +
theme_graph(background = grey(0.95), foreground = grey(0.5),
  plot_margin = margin(10, 10, 10, 10))
print(reh_pathways_plot)
```

```{r, eval = FALSE, include = FALSE}
svg("Ralstonia_H16_pathway_network.svg", width = 10, height = 8)
print(reh_pathways_plot)
dev.off()
```


----------

### Mapping genes expression to KEGG pathways

```{r}
# UTILITY FUNCTIONS
# -----------------

# remove other classes from graph object that are not required
# example_graph <- reh_pathways_as_graph$reh00020
# cleaned_graph <- graph::removeNode(
#   node = grep("ko:|cpd:|reh:", nodes(example_graph), value = TRUE), 
#   object = example_graph)
# 
# par(mfrow = c(2,1))
# plot(example_graph)
# plot(cleaned_graph)


# graph::subGraph(
#   snodes = grep("path:", nodes(reh_pathways_graph), value = TRUE), 
#   graph = reh_pathways_graph) %>% plot
#getKEGGedgeData(reh_pathways_graph) %>% slot("entry1ID")
```

